[tokens] Add DTCG-compliant design tokens package (1/2)

ryannono opened the pull request:
## Summary

Part 1 of 2: Adds the new `@blueprintjs/tokens` package implementing the W3C Design Tokens Community Group Format Module v2025.10 specification.

This PR establishes the foundation for Blueprint's migration to standardized design tokens using the DTCG format. Tokens are defined once in JSON and automatically generate **two independent outputs**: CSS custom properties (OKLCH colors) and SCSS variables (hex/rgba static values).

## Changes

### New Package: `@blueprintjs/tokens`

**Package Structure:**
- `src/tokens/base/`: Foundation tokens (color, dimension, typography, animation, layout)
- `src/tokens/semantic/`: Semantic tokens (intent, UI, shadows) with theme awareness
- `src/style-dictionary.ts`: Style Dictionary v5 configuration with custom transforms
- `src/build.ts`: Build script
- `dist/tokens.css`: Generated CSS custom properties with OKLCH colors
- `dist/tokens.scss`: Generated SCSS variables with static hex/rgba values

**Base Tokens** (src/tokens/base/):
- `color.tokens.json`: 72 colors in OKLCH color space with full light/dark scales
- `dimension.tokens.json`: Spacing, sizing, border widths
- `typography.tokens.json`: Font families, sizes, weights, line heights
- `animation.tokens.json`: Transition durations and easing functions
- `layout.tokens.json`: Z-index values

**Semantic Tokens** (src/tokens/semantic/):
- `intent.tokens.json`: Intent colors (primary, success, warning, danger)
- `ui.tokens.json`: UI semantic tokens (text, background, divider, focus, icon, interactive)
- `shadow.tokens.json`: Elevation shadows with proper layering

**Build System:**
- Style Dictionary v5 with custom transforms
- **Dual Distribution Strategy:** Independent CSS and SCSS outputs
  - **CSS Output**: OKLCH colors for perceptual uniformity, runtime theming
  - **SCSS Output**: Hex/rgba colors for compile-time compatibility, arithmetic support
- Custom name transform maps DTCG paths to Blueprint variable names (`color.blue.3` → `$blue3`)

**Generated Outputs:**

```scss
// dist/tokens.scss - Static SCSS values supporting arithmetic
$blue3: #2d72d2;
$pt-text-color: #1c2127;
$pt-spacing: 4px;
$pt-spacing * 2  // ✅ Arithmetic works!
```

```css
/* dist/tokens.css - CSS custom properties with OKLCH */
:root {
  --color-blue-3: oklch(0.5603 0.1634 257.63);
  --ui-text-default-light: oklch(0.2453 0.0136 253.08);
  --dimension-spacing-base: 4px;
}
```

## Key Features

- **DTCG v2025.10 Compliant**: Follows latest W3C specification
- **Dual Distribution Strategy**: CSS (OKLCH) and SCSS (hex/rgba) independent outputs
- **SCSS Arithmetic Support**: Static values enable `$pt-spacing * 2` operations
- **Zero Manual Maintenance**: Add token in JSON, get both CSS and SCSS automatically
- **Theme Support**: Light/dark theme variants
- **Type Safety**: Explicit `$type` declarations
- **Future-Ready**: CSS output prepared for runtime theming (Phase 2)

## Dual-System Architecture

The two outputs serve different purposes:
1. **SCSS version** (`tokens.scss`): Current Blueprint usage, backward compatibility, arithmetic operations
2. **CSS version** (`tokens.css`): Future CSS-only Blueprint distribution with runtime theming

This approach:
- ✅ Maintains backward compatibility (SCSS arithmetic still works)
- ✅ Avoids SCSS compile-time limitations with CSS variables
- ✅ Prepares foundation for future CSS-only distribution
- ✅ No breaking changes to existing Blueprint components

## Test Plan

- [ ] Install dependencies: `pnpm install`
- [ ] Build tokens: `cd packages/tokens && pnpm compile`
- [ ] Verify outputs exist: `packages/tokens/dist/tokens.css` and `packages/tokens/dist/tokens.scss`
- [ ] Check CSS uses OKLCH format (e.g., `oklch(0.5603 0.1634 257.63)`)
- [ ] Check SCSS uses hex/rgba format (e.g., `$blue3: #2d72d2`)
- [ ] Verify SCSS values are static (not `var()` references)
- [ ] Test SCSS arithmetic: `$pt-spacing * 2` should compile
- [ ] Review generated dist files in the PR diff

## Related

- Part 2: Core package integration (coming next)
- Implements W3C DTCG specification: https://www.designtokens.org/tr/2025.10/format/

Commit:
feat(tokens): Add DTCG-compliant design tokens package

Add new @blueprintjs/tokens package implementing W3C Design Tokens Community
Group Format Module v2025.10 specification.

Package structure:
- src/tokens/base/: Foundation tokens (color, dimension, typography, animation, layout)
- src/tokens/semantic/: Semantic tokens (intent, UI, shadows) with theme awareness
- src/style-dictionary.ts: Style Dictionary v4 configuration with custom transforms
- src/build.ts: Build script
- dist/: Generated CSS and SCSS outputs (committed for PR review visibility)

Key features:
- 72 colors in OKLCH color space with full light/dark scales
- Complete base tokens (color, dimension, typography, animation, layout)
- Semantic tokens (intent, UI, shadows) with theme awareness
- Style Dictionary v4 build pipeline with custom transforms
- CSS custom properties and SCSS variable outputs
- Generated dist files included for review visibility

Changes:
- Added @blueprintjs/tokens package
- Updated .gitignore to allow packages/tokens/dist/
- Updated pnpm-lock.yaml with new dependencies

Commit:
style(tokens): Format files with Prettier

Apply Prettier formatting to all tokens package files

Commit:
feat(tokens): Upgrade to Style Dictionary v5

- Update style-dictionary from ^4.0.1 to ^5.0.0
- v5 is a drop-in replacement with improved performance
- Build tested and working successfully

Commit:
docs(tokens): Update README to reflect Style Dictionary v5

svc-palantir-github commented:
<h3>docs(tokens): Update README to reflect Style Dictionary v5</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/fcad5c44-f808-4d8c-b3ec-8630b1eb75f6/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/fcad5c44-f808-4d8c-b3ec-8630b1eb75f6/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/fcad5c44-f808-4d8c-b3ec-8630b1eb75f6/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/fcad5c44-f808-4d8c-b3ec-8630b1eb75f6/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

Commit:
feat(tokens): Simplify to CSS variables only output

- Remove SCSS output generation (CSS variables only now)
- Update color transform to use hex/rgba instead of oklch for compatibility
- Simplify output path: dist/css/tokens.css → dist/tokens.css
- Update package exports to reflect new structure
- Remove SCSS-specific transforms and transform groups
- Update README documentation

svc-palantir-github commented:
<h3>feat(tokens): Simplify to CSS variables only output</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/f82b6f55-0767-4a92-8ea8-b903606dd594/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/f82b6f55-0767-4a92-8ea8-b903606dd594/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/f82b6f55-0767-4a92-8ea8-b903606dd594/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/f82b6f55-0767-4a92-8ea8-b903606dd594/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

Commit:
feat(tokens): Add auto-generated SCSS variable mappings

Implements Style Dictionary transforms and format to automatically generate
SCSS variables that map to CSS custom properties. This eliminates manual
maintenance and ensures SCSS and CSS stay perfectly in sync.

Key changes:
- Add custom name transform (name/scss-blueprint) to map DTCG tokens to
  Blueprint SCSS variable naming conventions
- Add custom value transform (value/css-var-reference) to convert all
  token values to var() references
- Add custom format (scss/variables-with-css-import) that generates
  tokens.scss with CSS import and variable mappings
- Add SCSS platform to Style Dictionary config
- Export tokens.scss in package.json

Generated output:
- dist/tokens.css: CSS custom properties with actual values
- dist/tokens.scss: SCSS variables mapped to CSS vars via var()

Example mappings:
  color.blue.3 → $blue3: var(--color-blue-3)
  ui.text.default.light → $pt-text-color: var(--ui-text-default-light)
  dimension.spacing.base → $pt-spacing: var(--dimension-spacing-base)

Benefits:
- Zero manual maintenance - add token in JSON, get SCSS mapping automatically
- Single source of truth - DTCG JSON drives everything
- Always in sync - can't get out of sync between CSS and SCSS
- Scalable - handles 200+ variables effortlessly

Commit:
docs(tokens): Update README for SCSS auto-generation

- Update Key Features to reflect hex/rgba output and SCSS mappings
- Update Directory Structure to show tokens.scss output
- Add SCSS Output Transforms section
- Expand Generated Output documentation
- Update usage examples to show SCSS var() references
- Add note about automatic CSS import in SCSS file

svc-palantir-github commented:
<h3>docs(tokens): Update README for SCSS auto-generation</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/59feb79d-a5b3-4f2c-af61-4911e5b75c03/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/59feb79d-a5b3-4f2c-af61-4911e5b75c03/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/59feb79d-a5b3-4f2c-af61-4911e5b75c03/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/59feb79d-a5b3-4f2c-af61-4911e5b75c03/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

Commit:
feat(tokens): Implement dual-system architecture with independent CSS/SCSS outputs

- CSS output: OKLCH colors for runtime theming and perceptual uniformity
- SCSS output: Static hex/rgba values supporting compile-time arithmetic
- Removes var() references from SCSS to enable $pt-spacing * 2 operations
- Updates README with dual-distribution strategy documentation
- Adds color/oklch transform for CSS, color/oklch-to-scss for SCSS
- Adds shadow/oklch and shadow/hex transforms for platform-specific outputs

This approach maintains backward compatibility while preparing for future
CSS-only Blueprint distribution.

Commit:
feat(tokens): Restore working Style Dictionary config with proper SCSS variable names

- Variable names now properly generated: $pt-intent-primary, $pt-heading-color, etc.
- SCSS output uses static values (SCSS variable references that resolve to hex/rgba)
- Fixed dist paths to output directly to dist/ instead of dist/css/ and dist/scss/
- Maintains backward compatibility with existing Blueprint SCSS code

svc-palantir-github commented:
<h3>feat(tokens): Restore working Style Dictionary config with proper SCSS variable names</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/d52a4535-d97e-4916-a07a-3165a1b1af93/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/d52a4535-d97e-4916-a07a-3165a1b1af93/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/d52a4535-d97e-4916-a07a-3165a1b1af93/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/d52a4535-d97e-4916-a07a-3165a1b1af93/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

Commit:
docs(tokens): Remove migration strategy phases section

svc-palantir-github commented:
<h3>docs(tokens): Remove migration strategy phases section</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/eff5e403-fb1b-4858-95c2-db791eeaa9da/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/eff5e403-fb1b-4858-95c2-db791eeaa9da/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/eff5e403-fb1b-4858-95c2-db791eeaa9da/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/eff5e403-fb1b-4858-95c2-db791eeaa9da/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

Commit:
refactor(core): Move tokens into core/src/common/tokens

- Move tokens package into core as src/common/tokens/
- Add token compilation to core build process
- Export tokens.css and tokens.scss from core package
- Tokens bundled with core distribution in lib/tokens/
- Remove separate tokens package

Commit:
fix(core): Update token paths and import CSS custom properties

- Fix source paths in style-dictionary (removed nested tokens dir)
- Import tokens.css in blueprint.scss for CSS custom property support
- Enables future migration from SCSS variables to CSS variables

Commit:
fix(core): Remove top-level await from token build script

- Convert async/await to promise chain for tsx compatibility
- Note: TypeScript compilation of token source files fails but tokens build successfully
- Token dist files are pre-generated and work correctly

Commit:
fix(core): Exclude token source from TypeScript compilation

- Token build scripts run with tsx, not compiled by TypeScript
- Fixes TS2307 errors about style-dictionary module

Commit:
feat(tokens): Align CSS variable naming with SCSS for seamless migration

Changes CSS custom property naming from DTCG format to Blueprint format to
enable seamless SCSS → CSS custom properties transformation.

Before:
- CSS: --color-blue-3 (DTCG naming)
- SCSS: $blue3 (Blueprint naming)
- Mismatch: var(--blue3) ≠ --color-blue-3 ❌

After:
- CSS: --blue3 (Blueprint naming)
- SCSS: $blue3 (Blueprint naming)
- Match: var(--blue3) = --blue3 ✅

Implementation:
- Added css/blueprint transform group using name/scss-blueprint transform
- Updated CSS platform to use css/blueprint instead of css/dtcg
- CSS variables now use OKLCH colors with Blueprint naming
- SCSS variables remain hex for backward compatibility
- Fixed source/buildPath to be relative to package root

This enables runtime theming while maintaining migration compatibility.
